tiledb/api/src/query/read/output/arrow.rs

```diff
@@ -1,258 +1,147 @@
 use std::sync::Arc;
 
 use arrow::array::{
-    Array as ArrowArray, FixedSizeListArray, GenericListArray, PrimitiveArray,
+    Array as ArrowArray, FixedSizeBinaryArray, FixedSizeListArray,
+    GenericListArray, LargeBinaryArray, PrimitiveArray,
 };
-use arrow::buffer::{OffsetBuffer, ScalarBuffer};
 use arrow::datatypes::Field;
 
 use crate::array::CellValNum;
 use crate::datatype::arrow::ArrowPrimitiveTypeNative;
-use crate::error::{DatatypeErrorKind, Error};
-use crate::query::buffer::{
-    Buffer, CellStructure, QueryBuffers, TypedQueryBuffers,
-};
+use crate::query::buffer::arrow::{Celled, QueryBufferArrowArray};
+use crate::query::buffer::TypedQueryBuffers;
 use crate::query::read::output::{RawReadOutput, TypedRawReadOutput};
-use crate::{typed_query_buffers_go, Result as TileDBResult};
+use crate::{typed_query_buffers_go, Datatype};
 
-impl<'data, C> TryFrom<RawReadOutput<'data, C>>
-    for PrimitiveArray<<C as ArrowPrimitiveTypeNative>::ArrowPrimitiveType>
+impl<C> TryFrom<RawReadOutput<'_, C>> for QueryBufferArrowArray<C>
 where
     C: ArrowPrimitiveTypeNative,
     PrimitiveArray<<C as ArrowPrimitiveTypeNative>::ArrowPrimitiveType>:
         From<Vec<C>> + From<Vec<Option<C>>>,
 {
-    type Error = crate::error::Error;
+    type Error = std::num::TryFromIntError;
 
-    fn try_from(value: RawReadOutput<C>) -> TileDBResult<Self> {
-        type MyPrimitiveArray<C> =
-            PrimitiveArray<<C as ArrowPrimitiveTypeNative>::ArrowPrimitiveType>;
+    fn try_from(value: RawReadOutput<C>) -> Result<Self, Self::Error> {
+        Self::try_from(Celled(value.ncells, value.input))
+    }
+}
 
-        match value.input.cell_structure {
-            CellStructure::Fixed(nz) if nz.get() == 1 => {}
-            structure => {
-                return Err(Error::Datatype(
-                    DatatypeErrorKind::UnexpectedCellStructure {
-                        context: None,
-                        expected: CellValNum::single(),
-                        found: structure.as_cell_val_num(),
-                    },
-                ))
-            }
+impl TryFrom<TypedRawReadOutput<'_>> for Arc<dyn ArrowArray> {
+    type Error = std::num::TryFromIntError;
+
+    fn try_from(value: TypedRawReadOutput<'_>) -> Result<Self, Self::Error> {
+        /*
+        fn assign_logical_type<PT, LT, I, O>(p: I) -> O where
+            PT: ArrowPrimitiveTypeNative,
+            LT: LogicalType<PhysicalType = PT> + ArrowPrimitiveTypeLogical,
+            I: PrimitiveArray<<<LT as LogicalType>::PhysicalType as ArrowPrimitiveTypeNative>::ArrowPrimitiveType>,
+            O: PrimitiveArray<<LT as ArrowPrimitiveTypeLogical>::ArrowPrimitiveType>
+        {
+            unimplemented!()
+        }
+        */
+
+        fn assign_logical_type<C>(
```

Comment:
This function is what we need because the `TypedQueryBuffers` enum variants are based on the physical type, not the logical type.

When we do `QueryBufferArrowArray::from(TypedQueryBuffers)`, the primitive arrays get a field whose data type is just the physical type - e.g. all of the date/time types will be Int64.  But this can't be used later on for the RecordBatch whose schema was constructed with a date/time field - it expects the corresponding primitive type, even though the fields have the same bit width.  (as much as this was annoying here, it's probably the right call because of possible expectations of the bit representation)

I think we do need the RecordBatch to have the logical type, not the physical type - that's the way we're going to communicate with DataFusion, after all, and Tables is certainly going to want to have date/time columns.  So I believe it is important to do something here, rather than change the schema that we give to the RecordBatch.

Anyway, re-typing the primitive array was a bit tricky to figure out, but this code seems to work.

Note for the review-ee, write some of this down in a comment for this inner function.